streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
+        final Map<String, String> usedClientTags = clientTagGetter.apply(usedClient);
+
+        for (final Map.Entry<String, String> usedClientTagEntry : usedClientTags.entrySet()) {
+            final String tagKey = usedClientTagEntry.getKey();
+
+            if (!rackAwareAssignmentTags.contains(tagKey)) {
+                LOG.warn("Client tag with key [{}] will be ignored when computing rack aware standby " +
+                         "task assignment because it is not part of the configured rack awareness [{}].",
+                    tagKey, rackAwareAssignmentTags);
+                continue;
+            }
+
+            final Set<String> allTagValues = tagKeyToValues.get(tagKey);
+
+            if (allTagValues.size() <= countOfUsedClients) {
+                allTagValues.forEach(tagValue -> tagEntryToUsedClients.remove(new KeyValue<>(tagKey, tagValue)));
+            } else {
+                final String tagValue = usedClientTagEntry.getValue();
+                final KeyValue<String, String> tagEntry = new KeyValue<>(tagKey, tagValue);
+                final Set<ProcessId> clientsOnUsedTagValue = tagEntryToClients.get(tagEntry);
+                tagEntryToUsedClients.put(tagEntry, clientsOnUsedTagValue);
+            }
+        }
+    }
+
+    private static Function<KafkaStreamsState, Map<String, String>> createClientTagGetter(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+        final boolean canPerformRackAwareOptimization = canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY);
+
+        if (hasRackAwareAssignmentTags || !canPerformRackAwareOptimization) {
+            return KafkaStreamsState::clientTags;
+        } else {
+            return state -> mkMap(mkEntry("rack", state.rackId().get()));
+        }
+    }
+
+    private static List<String> getRackAwareAssignmentTags(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+
+        if (hasRackAwareAssignmentTags) {
+            return applicationState.assignmentConfigs().rackAwareAssignmentTags();
+        } else if (canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY)) {
+            return Collections.singletonList("rack");
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    private static MoveStandbyTaskPredicate getStandbyTaskMovePredicate(final ApplicationState applicationState) {
+        final boolean hasRackAwareAssignmentTags = !applicationState.assignmentConfigs().rackAwareAssignmentTags().isEmpty();
+        final boolean canPerformRackAwareOptimization = canPerformRackAwareOptimization(applicationState, AssignedTask.Type.STANDBY);
+        if (hasRackAwareAssignmentTags || canPerformRackAwareOptimization) {
```

Comment:
Was thinking about this a bit and decided it should just be `if (hasRackAwareAssignmentTags)` -- for one thing, this method should only ever be invoked if `canPerformRackAwareOptimization` is true since it's called like halfway through `#optimizeRackAwareStandbyTasks`. 

For another, everything inside this `if` case is all about client tags, so it seems like `hasRackAwareAssignmentTags` must be true to enter this part of the code